[OpenMP][AMDGPU] Adapt dynamic callstack sizes to HIP behavior

mhalk opened the pull request:
Added a mechanism to cap values provided via LIBOMPTARGET_STACK_SIZE to a GFX-dependent value.

Changed several minor properties to be in sync with HIP:
  1. Default device stack size: 1024 / 1 KiB (hipLimitStackSize).
  2. During AQL packet generation in case of a dyn callstack the maximum between user-provided and compiler-default is chosen.
  3. Make sure we only allow 32bit values for stack size.

Added testcase where a dynamic stack is required due to recursion.

mhalk committed:
[OpenMP][AMDGPU] Adapt dynamic callstack sizes to HIP behavior

Added a mechanism to cap values provided via LIBOMPTARGET_STACK_SIZE to a
GFX-dependent value.

Changed several minor properties to be in sync with HIP:
  1. Default device stack size: 1024 / 1 KiB (hipLimitStackSize).
  2. During AQL packet generation in case of a dyn callstack the maximum
     between user-provided and compiler-default is chosen.
  3. Make sure we only allow 32bit values for stack size.

Added testcase where a dynamic stack is required due to recursion.

mhalk commented:
As stated in the testcase, I would have liked to add a failing configuration, which demonstrates the impact of insufficiently sized scratch memory -- but those cases result in a hang.
If anybody has suggestions / an opinion on this, I would be very grateful.
Even when it comes to removing this run configuration altogether.

jhuber6 commented:
@JonChesterfield was the one who originally encouraged me to leave the default at 16 KiB like it was before the COV5 change, so I'll leave that decision up to him.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1872,6 +1873,38 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+
+    // See: 'getMaxWaveScratchSize' in 'llvm/lib/Target/AMDGPU/GCNSubtarget.h'.
+    // But we need to divide by WavefrontSize.
+    if (GfxGen < 11) {
+      // 13-bit field in units of 256-dword.
+      MaxThreadScratchSize = ((256 * 4) / WavefrontSize) * ((1 << 13) - 1);
+    } else {
+      // 15-bit field in units of 64-dword.
+      MaxThreadScratchSize = ((64 * 4) / WavefrontSize) * ((1 << 15) - 1);
+    }
+
```
Can't decide if this should be a ternary or something. LLVM style normally omits braces on single line blocks.

mhalk replied:
Thanks & good point, I'll keep this in mind when addressing further feedback.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1872,6 +1873,38 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    // assume that AMDGPU offload archs are prefixed with "gfx" and suffixed
+    // with a two char arch specialization. In-between is the 1-2 char
+    // generation number we want to extract.
+    std::string CUKind{ComputeUnitKind};
```
This code looks really complicated and should be using `StringRef`. Are we just trying to get the major version? That should be pretty trivial.
```c++
StringRef Arch(ComputeUnitKind)
unsigned Version = 0u;
if (llvm::to_integer(Arch.drop_front(sizeof("gfx") - 1), Version))
  return Plugin::error("Invalid GFX architecture string");
unsigned Major = Version / 100;
```

mhalk replied:
Yes we only want the major version, so 9 for `gfx90a` or 10 for `gfx1030`.
It is pretty complicated for what it actually does. With the C++ tools I was aware of, I wanted to make sure I do not produce garbage.
Thank you very much!

jhuber6 replied:
Sorry, I forgot a `!` on that condition, the string conversion returns false on failure.

mhalk replied:
Thanks for the correction!
Also, e.g. "gfx90a" will fail, so I'd go for a slice:
```C++
    StringRef Arch(ComputeUnitKind);
    unsigned GfxGen = 0u;
    if (!llvm::to_integer(Arch.slice(sizeof("gfx") - 1, Arch.size() - 2),
                          GfxGen))
```
In any way, this is significantly easier to read.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1872,6 +1873,25 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    // But we need to divide by WavefrontSize.
+    // For generations pre-gfx11: use 13-bit field in units of 256-dword,
+    // otherwise: 15-bit field in units of 64-dword.
+    MaxThreadScratchSize = (GfxGen < 11)
```
if you get `gfx1100` the `GfxGen` will be `1100` here.

Also I don't know if I like `GfxGen` here. It's called either "Arch" or `ISAVersion` elsewhere that I know of.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1872,6 +1873,25 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    // generation number we want to extract.
+    StringRef Arch(ComputeUnitKind);
+    unsigned GfxGen = 0u;
+    if (!llvm::to_integer(Arch.slice(sizeof("gfx") - 1, Arch.size() - 2),
```
Why did the `drop_front` not work again? Also you need to specify that it's a hex radix.

jhuber6 commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -1872,6 +1873,25 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    MaxThreadScratchSize = (GfxGen < 11)
+                               ? ((256 * 4) / WavefrontSize) * ((1 << 13) - 1)
+                               : ((64 * 4) / WavefrontSize) * ((1 << 15) - 1);
+
```
That function is in a utility class, but could easily be made static or put somewhere more common. It uses an internal `generation` enum, but we could probably just convert to that here.

JonChesterfield commented:
I suggested we just pass too-large values to HSA and catch the error from HSA. Unfortunately it comes in the form of an invocation of callbackError which calls FATAL_MESSAGE which calls abort(), so that's not great. There's quite a lot of calls to FATAL_MESSAGE in the nextgen plugin infra which is a shame given the plugin put effort into returning error codes on other paths.

Leaning slightly towards accepting the copy&paste with string slice and just having things blow up on us later, what do you guys think?

JonChesterfield commented:
Make a header under llvm/lib/Target/AMDGPU and include it from the backend and from here. Provided it's an inline function in the header there's no messing around with cmake. The llvm libs are already in scope for openmp runtime. I'd say trivial technical complexity and moderate chance of the backend developers not liking the patch spawning moderate political complexity.

JonChesterfield commented:
@mhalk found what we think is the corresponding logic in hip at https://github.com/ROCm-Developer-Tools/clr/blob/develop/rocclr/device/device.cpp#L777, which in traditional copy&paste style has failed to notice that gfx11 is different to gfx10

b-sumner commented:
> @jayfoad @b-sumner any comments on moving the duplicated code above to a header in llvm/lib/Target/AMDGPU?

I don't think the HIP/OpenCL runtime wants this dependence.  So it would require a new API to be introduced in COMgr.  "The llvm libs are already in scope for openmp runtime" is surprising to me.

jhuber6 commented:
> > @jayfoad @b-sumner any comments on moving the duplicated code above to a header in llvm/lib/Target/AMDGPU?
> 
> I don't think the HIP/OpenCL runtime wants this dependence. So it would require a new API to be introduced in COMgr. "The llvm libs are already in scope for openmp runtime" is surprising to me.

The OpenMP libraries live in tree, so it's fairly trivial to include LLVM. OpenMP uses the LLVM backends for JIT functionality, as well as a load of ELF handling. We also use LLVM error handling and data structures almost everywhere. The LLVM libraries are generally statically linked unless shared libraries are explicitly requested by the user, so it introduces no extra dependencies in a release.

jdoerfert commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -3198,7 +3233,8 @@ Error AMDGPUKernelTy::launchImpl(GenericDeviceTy &GenericDevice,
   // Push the kernel launch into the stream.
   return Stream->pushKernelLaunch(*this, AllArgs, NumThreads, NumBlocks,
-                                  GroupSize, StackSize, ArgsMemoryManager);
+                                  GroupSize, static_cast<uint32_t>(StackSize),
```
Change it to uint32_t everywhere instead. Let's assume we stop at 4GB stack size.

mhalk replied:
Thanks -- Sure, sounds reasonable.

jdoerfert commented on openmp/libomptarget/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -2782,7 +2812,12 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
   /// not be statically determined.
-  uint64_t StackSize = 16 * 1024 /* 16 KB */;
+  /// Default: 1024, in conformity to hipLimitStackSize.
+  uint64_t StackSize = 1024 /* 1 KiB */;
```
uint32_t

mhalk commented:
> Jay's suggestion sounds good to me. HIP has different tradeoffs from living in a different repo - maybe the right thing to do there is include the header in comgr and re-export the function, but it's all very heavy handed relative to the #include and done approach in tree runtimes get to use.

On my first attempt to rename & move the corresponding `Generation` enum into `lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h` I would need to adapt (at least) ~ 25 other files within the backend; and TBH I've yet to get it to work as expected.

@jayfoad @JonChesterfield @jhuber6
Q: How should I go about this change to avoid introducing (lingering?) errors into the backend? Any suggestions welcome!
(Is there a way to narrow that number? [I guess not.])

jhuber6 commented:
> > Jay's suggestion sounds good to me. HIP has different tradeoffs from living in a different repo - maybe the right thing to do there is include the header in comgr and re-export the function, but it's all very heavy handed relative to the #include and done approach in tree runtimes get to use.
> 
> On my first attempt to rename & move the corresponding `Generation` enum into `lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h` I would need to adapt (at least) ~ 25 other files within the backend; and TBH I've yet to get it to work as expected.
> 
> @jayfoad @JonChesterfield @jhuber6 Q: How should I go about this change to avoid introducing (lingering?) errors into the backend? Any suggestions welcome! (Is there a way to narrow that number? [I guess not.])

Can you just make an inline function somewhere that the current implementation also calls?

mhalk commented:
> > > Jay's suggestion sounds good to me. HIP has different tradeoffs from living in a different repo - maybe the right thing to do there is include the header in comgr and re-export the function, but it's all very heavy handed relative to the #include and done approach in tree runtimes get to use.
> > 
> > 
> > On my first attempt to rename & move the corresponding `Generation` enum into `lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.h` I would need to adapt (at least) ~ 25 other files within the backend; and TBH I've yet to get it to work as expected.
> > @jayfoad @JonChesterfield @jhuber6 Q: How should I go about this change to avoid introducing (lingering?) errors into the backend? Any suggestions welcome! (Is there a way to narrow that number? [I guess not.])
> 
> Can you just make an inline function somewhere that the current implementation also calls?

Another level of indirection sounds good to me, I'll check that.

mhalk commented:
FYI: Just a quick update, as discussed offline, I'll first work on solely moving / extracting the `Generation` enum with the intent to do this in its own PR (as a precursor to this one). Once it's ready, I'll link it here.

arsenm commented:
what's the status of this?

jhuber6 commented:
Status? We should probably try to get this in before the release if possible.

arsenm reviewed (changes requested):
needs resolve to main

mhalk commented:
Apologies for the long radio silence.

After reaching out to @arsenm the idea of moving the enum `AMDGPUSubtarget::Generation` does not seem attractive anymore.
As Matt suggested I'll try to implement a `SubtargetFeature` (named sth. like `MaxWaveScratchSize`) that may be queried via `TargetParser`, which holds the corresponding callstack size.
This will avoid duplication, if I can get it to work within the two locations where `getMaxWaveScratchSize` is needed.
(These changes should be reasonable within the same PR.)

I plan to be working on this next week.

mhalk commented:
When starting to actually try and implement this, I came to the conclusion that when adding such a 'SubtargetFeature' it would really just duplicate `GCNSubtarget::getMaxWaveScratchSize`.
(IMHO this does not seem like a desirable / acceptable solution when looking at the former discussion.)

On top of that, TBH I only see `bool` features that may be queried via `TargetParser` but we want to encode an `unsigned` value. (So, if that is in fact a desired solution: how would one create such an unsigned feature?)

If we would really want to avoid duplication IMO (maybe I'm wrong) we would have to either:
 * Use `GCNSubtarget::getMaxWaveScratchSize` (somehow?) at the particular location within the OpenMP amdgpu plugin (I have the Triple available, so theoretically I have the information I need).
 * Encode an `unsigned` value which holds the `MaxWaveScratchSize` and can be queried via `TargetParser` at both locations where this info is needed (OpenMP amdgpu plugin and GCNSubtarget class).

Any suggestions or ideas are very welcome & appreciated.

arsenm commented:
> On top of that, TBH I only see `bool` features that may be queried via `TargetParser` but we want to encode an `unsigned` value. (So, if that is in fact a desired solution: how would one create such an unsigned feature? Or are there any examples I might have missed?)

You can encode arbitrary values in a subtarget feature. The current main example would be FeatureLocalMemorySize*. Not sure if we directly expose those in the TargetParser

mhalk commented:
> You can encode arbitrary values in a subtarget feature. The current main example would be FeatureLocalMemorySize*. Not sure if we directly expose those in the TargetParser

Great, thank you @arsenm! That helped a lot.

Now I've added the corresponding three `SubtargetFeatures` and added the getters similar to `(Feature)LocalMemorySize`.
TBH it is still not entirely clear to me how to access the newly available `(get)MaxWaveScratchSize ` via the `TargetParser` or `AMDGPUBaseInfo`.

When using the `TargetParser` I was able to construct the FeatureMap corresponding to my present device(s) but that did not contain the info (and from what it contained, MaxWaveScratchSize does not really seem to 'fit' in that group of device capabilities).

Do you have another pointer for me on how to proceed?

mhalk commented:
> The TargetParser does seem to only be exposing boolean-ish looking names, which unfortunately would imply introducing N different feature names there. I've lost track of why you need this in the first place; if you over-commit on the stack size, won't the lower level API fail for you?

When launching a kernel which requires a dynamic callstack we want to use the maximum target-specific `StackSize`.

Yes, such a fail will occur, but IIRC this is not desired:
```
AMDGPU fatal error 1: Received error in queue 0x7f24b2a44000:
HSA_STATUS_ERROR_OUT_OF_RESOURCES: The runtime failed to allocate the necessary resources.
This error may also occur when the core runtime library needs to spawn threads or create internal OS-specific events.
Aborted (core dumped)
```
(IMHO in a complex scenario this message might not be very helpful.)

With the current changes the user will be informed, without aborting and while using the max. scratch memory.
(Unfortunately, it's still a matter of relaying the info into the OpenMP amdgpu plugin, so basically the situation hasn't changed much.)

arsenm commented:
> Yes, such a fail will occur, but IIRC this is not desired:

I think it would be better if we fixed the low level diagnostics to not be garbage, rather than building another layer of diagnostics on top
